app/controllers/bridges/oai/oai_controller.rb

```diff
@@ -200,41 +235,68 @@ def filter_tasks_by_set(tasks, set_spec)
         tasks.where(id: TaskLabel.select(:task_id).where(label_id: slices[2].to_i))
       end
 
-      def valid_set_spec?(set_spec)
-        set_spec.downcase.match?(/^task$|^task:label$|^task:label:[0-9]+$/)
+      def filter_tasks_by_resumption_params(tasks, resumption_params)
+        tasks.where('updated_at < :ts_until AND (updated_at > :ts_from OR (updated_at = :ts_from AND id > :last_id))', resumption_params)
       end
 
-      def oai_params
-        params.permit(:verb, :identifier, :from, :until, :metadataPrefix, :set)
+      def valid_set_spec?(set_spec)
+        set_spec.downcase.match?(/^task$|^task:label$|^task:label:[0-9]+$/)
       end
 
-      def parse_time_bounds!(params)
-        tfrom = nil
-        tuntil = nil
-
-        if params[:from]
+      def parse_parameters!(params)
+        if params.key? :resumptionToken
           begin
-            tfrom = DateTime.iso8601(params[:from])
-          rescue Date::Error
-            raise OaiError.new("The specified time '#{params[:from]}' is not in iso8601 format", 'badArgument')
+            params = ActionController::Parameters.new(JSON.parse(Base64.decode64(params[:resumptionToken])))
```

Comment:
Similar to above, we should probably parse without linefeeds and URL safe:

```suggestion
            params = ActionController::Parameters.new(JSON.parse(Base64.urlsafe_decode64(params[:resumptionToken])))
```